discord/http.py

```diff
@@ -260,123 +235,126 @@ async def request(
-                            if is_global:
-                                _log.warning(
-                                    (
-                                        "Global rate limit has been hit. Retrying in"
-                                        " %.2f seconds."
-                                    ),
-                                    retry_after,
+                            if remaining:
+                                remaining = int(remaining)
+                            if reset_after:
+                                reset_after = float(reset_after)
+
+                            bucket.set_metadata(remaining, reset_after)
+
+                            # the request was successful so just return the text/json
+                            if 300 > response.status >= 200:
+                                _log.debug("%s %s has received %s", method, url, data)
+                                return data
+
+                            # we are being rate limited
+                            if response.status == 429:
+                                if not response.headers.get("Via") or isinstance(
+                                    data, str
+                                ):
+                                    # Banned by Cloudflare more than likely.
+                                    raise HTTPException(response, data)
+
+                                fmt = (
+                                    "We are being rate limited. Retrying in %.2f seconds."
+                                    ' Handled under the bucket "%s"'
                                 )
-                                self._global_over.clear()
-
-                            await asyncio.sleep(retry_after)
-                            _log.debug("Done sleeping for the rate limit. Retrying...")
 
-                            # release the global lock now that the
-                            # global rate limit has passed
-                            if is_global:
-                                self._global_over.set()
-                                _log.debug("Global rate limit is now over.")
+                                # sleep a bit
+                                retry_after: float = data["retry_after"]
+                                _log.warning(fmt, retry_after, bucket_id)
+                                is_global: bool = data.get("global", False)
+
+                                if is_global:
+                                    self.global_dynamo = DynamicBucket()
+                                    await self.global_dynamo.executed(
+                                        retry_after,
+                                        remaining or 10,
+                                        is_global=is_global,
+                                    )
+                                    self.global_dynamo = None
+                                else:
+                                    tbucket = DynamicBucket()
```

Comment:
I don't understand why dynamic buckets are used here. Shouldn't permanent buckets be used instead?

Reply:
permanent buckets are for buckets which we know the rate limit of and can predict, dynamic buckets are for when a permanent bucket misses a prediction or if the limit of a permanent bucket is unknown.